### PR TITLE
PDF text extractor

### PR DESCRIPTION
**Pull request by theUnkownName**
Extracts text from PDF.
All images will be discarded

### CODE REVIEW

File: crafters/nlp/PDFExtractor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    :class:`PDFTextExtractor` Extracts text from PDF.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+    def craft(self, uri: str, *args, **kwargs):
+        import PyPDF2
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:18]: move the import to `post_init`

File: crafters/nlp/PDFExtractor/__init__.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    """
+    :class:`PDFTextExtractor` Extracts text from PDF.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+    def craft(self, uri: str, *args, **kwargs):
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:17]: what if the input is not a local file but it comes as `bytes`?

File: crafters/nlp/PDFExtractor/__init__.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        import PyPDF2
+        pdf_obj = open(uri, 'rb')
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:19]: this file needs to be also properly closed

File: crafters/nlp/PDFExtractor/__init__.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        pdf_reader = PyPDF2.PdfFileReader(pdf_obj)
+        page_obj = pdf_reader.getPage(0)
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:21]: why do we just take first page? The library offers the `read(self, stream)` interface that I think takes all text

File: crafters/nlp/PDFExtractor/tests/test_pdfextractor.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                              "In posuere auctor pellentesque. Donec cursus ipsum ut odio dictum ultricies. "
+                              "Suspendisse volutpat accumsan imperdiet. Pellentesque sodales pharetra egestas. S"
+                              "ed tincidunt nec eros vel mollis. Etiam blandit leo vitae mattis bibendum. "
+                              "Nunc rutrum hendrerit ligula. ")
+    canvas.save()
+
+def test_read_pdf():
+    crafter = PDFTextExtractor()
+    text = crafter.craft('cats_are_awesome.pdf')
+    print(text)
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:24]: u can have the text in a static variable and check the extracted value

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: crafters/nlp/PDFExtractor/__init__.py (hunk 4 above)
Comment (theUnkownName) [reply crafters/nlp/PDFExtractor/__init__.py:21]: Right! I missed this

File: crafters/nlp/PDFExtractor/tests/test_pdfextractor.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+from .. import PDFTextExtractor
+
+def create_pdf():
+    from reportlab.pdfgen.canvas import Canvas
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:4]: if you use the page interface, make sure you are creating a pdf with multiple pages

Comment (theUnkownName) [reply crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:4]: Yeah the example one has 2 pages with one image to test that

File: crafters/nlp/PDFExtractor/tests/test_pdfextractor.py [hunk 7]
```diff
@@ -1,25 +1,18 @@
...
     crafter = PDFTextExtractor()
-    text = crafter.craft('cats_are_awesome.pdf')
+    text = crafter.craft(uri='cats_are_awesome.pdf', buffer=None)
     print(text)
 
+
+def test_io_buffer():
+    import requests
+    url = 'http://ceur-ws.org/Vol-2410/paper35.pdf'
+    response = requests.get(url)
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:14]: Download or create a complicated PDF (open source one if downloaded), But do not add requests in our test.

1. Won't work if no connection available.
2. What if they remove this pdf from the site? We would not have a clue what's wrong

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: crafters/nlp/PDFExtractor/__init__.py [hunk 8]
```diff
@@ -14,11 +14,10 @@ def __init__(self, *args, **kwargs):
 
     def craft(self, uri: str, buffer: bytes, *args, **kwargs):
         import PyPDF2
-        import io
 
         text = ""
         if buffer:
-            pdf_obj = io.BytesIO(buffer)
+            pdf_obj = buffer
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:20]: buffer object has no `close`. PdfFileReader accepts a `stream` that is what `io.BytesIO` provides.

File: crafters/nlp/PDFExtractor/tests/test_pdfextractor.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+
+
+def test_io_uri():
+    crafter = PDFTextExtractor()
+    text = crafter.craft(uri='cats_are_awesome.pdf', buffer=None)
+    print(text)
+
+
+def test_io_buffer():
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:11]:
```suggestion
def test_io_buffer():
    with open('paper35.pdf', 'rb') as fp:
        input_bytes = fp.read()
        
     crafter = PDFTextExtractor()
     text = crafter.craft(uri=None, buffer=input_bytes)
     assert text == expected_text
```

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: crafters/nlp/PDFExtractor/__init__.py [hunk 10]
```diff
@@ -12,23 +12,29 @@ class PDFTextExtractor(BaseCrafter):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-    def craft(self, uri: str, buffer: bytes, *args, **kwargs):
+    def post_init(self):
         import PyPDF2
+        import io
 
+    def craft(self, uri: str, buffer: bytes, *args, **kwargs):
+        import PyPDF2
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:20]: remove these imports then

File: crafters/nlp/PDFExtractor/__init__.py [hunk 11]
```diff
@@ -12,23 +12,29 @@ class PDFTextExtractor(BaseCrafter):
...
-            page = pdf_reader.getPage(i)
-            text += page.extractText()
+        try:
+            pdf_reader = PyPDF2.PdfFileReader(pdf_obj)
+            count = pdf_reader.numPages
+            for i in range(count):
+                page = pdf_reader.getPage(i)
+                text += page.extractText()
+        except IOError:
+            print("Could not open file")
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/__init__.py:37]: do not handle `exception` here, it is handled by `jina` at different levels depending on configurations.

File: crafters/nlp/PDFExtractor/tests/test_pdfextractor.py [hunk 12]
```diff
@@ -1,15 +1,20 @@
 from .. import PDFTextExtractor
 
 
+expected_text = "A cat poem\nI love cats, I love every kind of cat,\nI just wanna hug all of them, but I can't," \
```
Comment (JoanFM) [inline crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:4]: Perfect, how does it then behave with pdfs that are in chinese?

Comment (theUnkownName) [reply crafters/nlp/PDFExtractor/tests/test_pdfextractor.py:4]: Looks like it's not encoding those characters, they are being ignored at the moment

**Pull request closed without merging**